### PR TITLE
feat: validator retry, backoff, and rate limiting (LAB-6637)

### PR DESCRIPTION
**Pull request by NextG3nMxJ**
## Summary

Infrastructure resilience for the validator engine. Validation and scoring stay programmatic — titus does not call an LLM.

- Adaptive backoff controller with jitter
- `RetryHTTPClient` for 429/5xx/connection errors (`Retry-After` capped)
- Wired into YAML HTTP validators and Go validators that own an `*http.Client`
- `--validate-rate-limit` on `titus scan` (also applied to `ValidateMatch`)

LLM verifier, `pkg/llm`, YAML `llm:` scoring conditions, `ResponseMeta`, and the spec/plan docs have been removed from this PR.

## Test plan

- [x] `go test ./pkg/validator ./pkg/scoring ./pkg/types ./cmd/titus`
- [ ] CI green

### CODE REVIEW

File: pkg/validator/engine.go [hunk 1]
```diff
@@ -53,7 +54,9 @@ type Engine struct {
 	validators []Validator
 	cache      *ValidationCache
 	workers    int
-	sem        chan struct{} // semaphore for bounded concurrency
+	sem        chan struct{}      // semaphore for bounded concurrency
+	limiter    *rate.Limiter      // optional rate limiter
+	backoff    *BackoffController // optional adaptive backoff
```
Comment (NextG3nMxJ) [inline pkg/validator/engine.go]: **[bug]** Phase 1 retry/backoff is dead in production. `initValidationEngine` always constructs a `BackoffController` and calls `SetBackoff`, but `Engine` never reads `e.backoff` in `ValidateAsync`/`ValidateMatch` (only the new limiter is consulted). `RetryHTTPClient` is not referenced outside `retry.go` / `retry_test.go`, and `HTTPValidator` / Go validators still call `*http.Client.Do` directly. `TestEngine_WithBackoff` only asserts the pointer is non-nil, so CI cannot catch that validation HTTP still has no retry, no jittered backoff, and no consecutive-error tracking.

**Suggestion:** Either plumb a shared `RetryHTTPClient` (fed by `e.backoff`) into `HTTPValidator` and the Go validators, or stop constructing/setting unused backoff in `initValidationEngine`. If wiring is deferred, drop the CLI-side `SetBackoff` so the engine does not advertise resilience it does not provide.

File: pkg/types/validation.go [hunk 2]
```diff
@@ -14,11 +14,21 @@ const (
...
-	Confidence  float64           `json:"confidence"`
-	Message     string            `json:"message"`
-	ValidatedAt time.Time         `json:"validated_at"`
-	Details     map[string]string `json:"details,omitempty"` // Extended validation details
+	Status       ValidationStatus  `json:"status"`
+	Confidence   float64           `json:"confidence"`
+	Message      string            `json:"message"`
+	ValidatedAt  time.Time         `json:"validated_at"`
+	Details      map[string]string `json:"details,omitempty"` // Extended validation details
+	ResponseMeta *ResponseMeta     `json:"response_meta,omitempty"`
```
Comment (NextG3nMxJ) [inline pkg/types/validation.go]: **[bug]** `ResponseMeta` is JSON-tagged and attached on every HTTP/Jenkins validation, not only when `--llm-verify` is on. `titus scan --format=json` encodes `[]*types.Match` (and findings embed those matches), so `--validate` now emits up to 2KB of upstream response body (base64) plus the request URL on every result. URL templates and echoed API bodies can contain the secret or other credential material. SQLite persistence is unchanged (status/message only), but JSON/library marshal paths are a new leak with LLM features off.

**Suggestion:** Tag `ResponseMeta` with `json:"-"` (keep it in-memory for the verifier), or populate it only when an LLMVerifier is wrapping the engine. Do not serialize raw bodies/URLs in scan JSON.

File: pkg/llm/sanitize.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	}
+	return string(body[:maxLen]) + "... [truncated]"
+}
+
+// WrapUntrusted sanitizes content and wraps it in an XML-style tag, to
+// clearly delineate untrusted input within an LLM prompt as a mitigation
+// against prompt injection.
+func WrapUntrusted(tag, content string) string {
+	content = Sanitize(content)
+	return fmt.Sprintf("<%s>\n%s\n</%s>", tag, content, tag)
```
Comment (NextG3nMxJ) [inline pkg/llm/sanitize.go]: **[bug]** `WrapUntrusted` is the stated prompt-injection mitigation but it does not escape `</tag>` (or the tag name) inside content. An HTTP body or secret containing `</response_body>` / `</scoring_prompt>` breaks the wrapper. That is reachable: Jenkins validation uses a snippet-derived URL (`extractJenkinsURL`), so an untrusted repo can point validation at an attacker-controlled host whose body is then forwarded to the LLM; `llmCondition` also wraps attacker-controlled `{{secret}}` / named groups. A successful breakout can flip undetermined → valid (verifier) or force `fires_when` (scoring).

**Suggestion:** Escape `<`/`>` (or use a delimiter the model is told is opaque, e.g. a random fence) before wrapping. Add tests for embedded closing tags. Treat user-controlled validation URLs as untrusted input in the verifier prompt, not just the body.

File: pkg/validator/retry.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+			if c.backoff != nil {
+				c.backoff.RecordSuccess()
+			}
+			return resp, nil
+		}
+	}
+
+	if lastResp != nil {
+		if c.backoff != nil {
+			c.backoff.RecordSuccess()
```
Comment (NextG3nMxJ) [inline pkg/validator/retry.go]: **[bug]** After retries are exhausted on 429/5xx, `Do` still calls `backoff.RecordSuccess()` because `lastResp != nil`. Consecutive-error state is therefore reset after a persistent failure, which is the opposite of adaptive backoff. The 5xx-persists test uses a nil backoff, so this path is untested.

**Suggestion:** Record success only for non-retryable final statuses (2xx/3xx/4xx-other-than-429). On exhausted 429/5xx, leave the error counter in place (or `RecordError` once) and return the last response.

File: pkg/validator/retry.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+	var lastErr error
+
+	for attempt := 0; attempt < retryMaxAttempts; attempt++ {
+		if attempt > 0 && lastResp != nil {
+			_, _ = io.Copy(io.Discard, lastResp.Body)
+			_ = lastResp.Body.Close()
+			lastResp = nil
+		}
+
+		resp, err := c.inner.Do(req)
```
Comment (NextG3nMxJ) [inline pkg/validator/retry.go]: **[suggestion]** `Do` reuses the same `*http.Request` on retry without rewinding `Body` / setting `GetBody`. GET tests pass; POST/PUT validators (and any YAML validator with a body) would send an empty body on the retry attempt once this client is wired. Connection-error retries have the same problem.

**Suggestion:** If `req.GetBody` is set, reset `req.Body` before each attempt; otherwise skip retries for requests with a non-nil, non-rewindable body (or wrap bodies with `GetBody` at construction). Add a POST-retry test.

File: pkg/validator/verifier.go [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+	if original.Status == types.StatusValid && newStatus == types.StatusInvalid {
+		return original
+	}
+	if original.Status == types.StatusInvalid && newStatus == types.StatusValid {
+		return original
+	}
+
+	return &types.ValidationResult{
+		Status:       newStatus,
+		Confidence:   verdict.Confidence,
```
Comment (NextG3nMxJ) [inline pkg/validator/verifier.go]: **[suggestion]** Never-downgrade only blocks valid↔invalid status flips. `applyLLMResponse` otherwise overwrites `Confidence` with the model's number, unclamped and with no floor. An undetermined result can become `valid` at confidence 0.0, and a low-confidence `valid` can have its confidence lowered while keeping status. LAB-6639 describes upgrading undetermined results, not blindly trusting arbitrary confidence.

**Suggestion:** Require a minimum confidence (e.g. ≥ 0.7) before accepting an LLM status change; clamp to `[0,1]`; when original status is already valid/invalid, keep `max(original.Confidence, llm)` (or skip the call entirely unless status is undetermined).

File: pkg/validator/verifier.go [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+func (v *LLMVerifier) ValidateAsync(ctx context.Context, match *types.Match) <-chan *types.ValidationResult {
+	engineCh := v.engine.ValidateAsync(ctx, match)
+	out := make(chan *types.ValidationResult, 1)
+	go func() {
+		defer close(out)
+		result := <-engineCh
+		if result == nil {
+			return
+		}
+		if !v.shouldCallLLM(result) || v.spent.Load() >= v.budget {
```
Comment (NextG3nMxJ) [inline pkg/validator/verifier.go]: **[suggestion]** Budget is checked with `spent.Load() >= budget` then later `spent.Add(1)` inside `tryLLMUpgrade`, after the semaphore is acquired. `ValidateAsync` (the CLI path) can run up to 4 LLM calls concurrently, so the cap can overshoot by the semaphore width. Cache hits correctly skip `spent`, but the TOCTOU is real under load.

**Suggestion:** Use `CompareAndSwap` / `Add` then roll back if the new value exceeds `budget`, or reserve the slot before calling `Complete`.

File: pkg/validator/verifier.go [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+	upgraded := v.tryLLMUpgrade(ctx, match, result)
+	if upgraded != result {
+		v.upgrades.Add(1)
+	}
+	return upgraded, nil
+}
+
+// ValidateAsync submits match to the wrapped engine's async path, then
+// applies the LLM second-pass review on the result when applicable.
+func (v *LLMVerifier) ValidateAsync(ctx context.Context, match *types.Match) <-chan *types.ValidationResult {
```
Comment (NextG3nMxJ) [inline pkg/validator/verifier.go]: **[suggestion]** All verifier unit tests and the integration test go through `ValidateMatch`. The scan CLI uses `ValidateAsync`, which has extra control flow (nil result closes the channel without sending; budget check is racy with the goroutine). A nil result would panic in verbose `validateMatches` (`result.Status`). There is no direct coverage of this path, matching the author's follow-up note.

**Suggestion:** Add `ValidateAsync` tests for upgrade, budget, skip-without-meta, context cancel, and the nil-result path (always send a result, never close empty).

File: pkg/scoring/condition_llm.go [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+	resp, err := c.client.Complete(ctx, &llm.Request{
+		System:    "You are analyzing a detected secret. Answer concisely.",
+		Messages:  []llm.Message{{Role: "user", Content: llm.WrapUntrusted("scoring_prompt", prompt)}},
+		MaxTokens: 256,
+	})
+	if err != nil {
+		return false, nil
+	}
+
+	return strings.Contains(strings.ToLower(resp.Content), strings.ToLower(c.firesWhen)), nil
```
Comment (NextG3nMxJ) [inline pkg/scoring/condition_llm.go]: **[suggestion]** `llmCondition` fires on a case-insensitive substring (`strings.Contains`), so `fires_when: admin` matches `administrator` / `not_admin`. There is no per-scan LLM budget or cache on this path (unlike the verifier), so `--score-scope` plus any `llm:` modifier can make one Anthropic call per match. `Evaluate` also sends substituted secret values to the provider (by design for `{{secret}}`) with no CLI warning. No builtin YAML currently uses `llm:`, so this is latent until custom scorers land.

**Suggestion:** Match `fires_when` as a token/exact response (or structured JSON), share the verifier budget/cache, and print a stderr warning that LLM scoring uploads secret material to the configured provider.

File: pkg/scoring/condition_llm.go [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+// substituteMatchVars replaces {{name}} and {{ name }} placeholders in tpl
+// with values drawn from the match: rule_name (RuleID), matching
+// (Snippet.Matching), and each entry in NamedGroups. Untrusted match data is
```
Comment (NextG3nMxJ) [inline pkg/scoring/condition_llm.go]: **[suggestion]** Comment says untrusted match data is "substituted verbatim" and that callers must sanitize, but `substituteMatchVars` already calls `llm.Sanitize` on every substitution. That is stale design-plan text and will mislead the next implementer.

**Suggestion:** Delete the verbatim/caller-must-sanitize sentence; state that values are sanitized before substitution and then XML-wrapped by `Evaluate`.

File: pkg/scoring/loader.go [hunk 11]
```diff
@@ -148,8 +160,10 @@ func convertYAMLScorer(ys yamlScorer) (*Scorer, error) {
 }
 
 // convertYAMLModifier enforces the "exactly one condition, exactly one action"
-// rule and compiles the regex (if any).
-func convertYAMLModifier(ym yamlModifier) (Modifier, error) {
+// rule and compiles the regex (if any). llmClient is injected by the caller
+// for llm: conditions; it may be nil when no LLM condition is present, or
+// when LLM scoring hasn't been wired up yet (see Task 14).
```
Comment (NextG3nMxJ) [inline pkg/scoring/loader.go]: **[suggestion]** `convertYAMLModifier` still comments "when LLM scoring hasn't been wired up yet (see Task 14)". That is plan-checklist residue; scoring is wired through `WithLLMClient` / `AllBuiltinScorers` in this PR.

**Suggestion:** Remove the Task 14 reference; note that a nil client makes `llm:` conditions evaluate false.

**Comment by NextG3nMxJ** [comment]
Review follow-up is in 56f9dd5.

**Codex / Claude / Gemini review bodies**

- Retry/backoff is no longer dead: `SetBackoff` wraps YAML `HTTPValidator` and the Go validators that own an `*http.Client`. `ValidateMatch` now also honors `--validate-rate-limit`.
- `ResponseMeta` is `json:"-"` so `--format=json` no longer emits bodies/URLs.
- `WrapUntrusted` HTML-escapes untrusted content. Scoring wraps match variables, not the trusted prompt template.
- LLM budget is reserved atomically; cache hits still apply after the cap.
- `--llm-timeout` bounds the whole Anthropic `Complete` call; retry sleeps only when another attempt remains.
- `AllBuiltinScorers()` is variadic again for library callers.
- `--llm-verify` / `--score-scope` print stderr warnings that secret-bearing data is sent to the LLM provider.
- `ValidateAsync` is unit-tested and always emits a result.

**Not changing**

- gosec SSRF on the retry wrapper: validators are supposed to hit caller-supplied URLs.
- Requiring TLS for `httptest` servers: loopback HTTP remains allowed for tests.

File: pkg/validator/backoff.go [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+// Wait blocks for the current backoff delay (if any), jittered, or until ctx
+// is done, whichever comes first. It returns nil if the delay elapsed (or no
+// delay was needed), or ctx.Err() if the context was cancelled first.
+func (b *BackoffController) Wait(ctx context.Context) error {
+	d := b.currentDelay()
+	if d == 0 {
+		return nil
+	}
+	jitter := time.Duration(rand.Int63n(int64(d)))
```
[the remark this answers is not in the document]
Comment (NextG3nMxJ) [reply pkg/validator/backoff.go]: Added `#nosec G404` on the jitter call. This is backoff delay jitter, not a cryptographic secret; same justification as `pkg/enum/gdrive.go` / `pkg/enum/asana.go`. `crypto/rand` is unnecessary here.

File: pkg/validator/retry.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (NextG3nMxJ) [reply pkg/validator/retry.go]: This wrapper replays the request the validator already built. Secret validation is supposed to contact caller-supplied endpoints, so this is not a new SSRF surface. Left the behavior as-is.

File: pkg/validator/verifier.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (NextG3nMxJ) [reply pkg/validator/verifier.go]: Fixed in 56f9dd5. Budget is reserved with `CompareAndSwap` after the cache lookup inside `tryLLMUpgrade`, so concurrent `ValidateAsync` callers cannot overshoot `--llm-budget`. Cache hits skip reservation and still apply after the cap is reached.

File: pkg/llm/anthropic.go [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+				if delay > 30*time.Second {
+					delay = 30 * time.Second
+				}
+			}
+		}
+		if delay > 0 {
+			t := time.NewTimer(delay)
+			defer t.Stop()
+			select {
+			case <-t.C:
```
[the remark this answers is not in the document]
Comment (NextG3nMxJ) [reply pkg/llm/anthropic.go]: Fixed in 56f9dd5. `Complete` now applies `--llm-timeout` to the whole operation via `context.WithTimeout` when the caller context has no deadline, and `Retry-After` sleeps only when another attempt remains.

File: pkg/validator/retry.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (NextG3nMxJ) [reply pkg/validator/retry.go]: Fixed in 56f9dd5. Before a retry, `Do` rewinds `req.Body` through `req.GetBody`. If the body is not replayable, it does not retry.

File: pkg/validator/retry.go [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+				c.backoff.RecordSuccess()
+			}
+			return resp, nil
+		}
+	}
+
+	if lastResp != nil {
+		if c.backoff != nil {
+			c.backoff.RecordSuccess()
+		}
```
[the remark this answers is not in the document]
Comment (NextG3nMxJ) [reply pkg/validator/retry.go]: Fixed in 56f9dd5. Exhausted 429/5xx returns the last response without `RecordSuccess`, so consecutive errors keep accumulating during a persistent outage.

File: pkg/llm/anthropic_test.go [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+func TestAnthropicClient_Success(t *testing.T) {
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, "POST", r.Method)
+		assert.Equal(t, "/v1/messages", r.URL.Path)
+		assert.Equal(t, "test-key", r.Header.Get("x-api-key"))
+		assert.Equal(t, "2023-06-01", r.Header.Get("anthropic-version"))
+
+		body, _ := io.ReadAll(r.Body)
+		var reqBody map[string]any
+		json.Unmarshal(body, &reqBody)
```
[the remark this answers is not in the document]
Comment (NextG3nMxJ) [reply pkg/llm/anthropic_test.go]: Fixed in 56f9dd5. Anthropic test handlers now check `json.Unmarshal` / `Encode` / `Write` errors.

File: pkg/llm/anthropic.go [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+	baseURL string
+	client  *http.Client
+	retries int
+}
+
+func newAnthropicClient(apiKey, model string, cfg clientConfig) (Client, error) {
+	baseURL := cfg.BaseURL
+	if baseURL == "" {
+		baseURL = defaultAnthropicURL
+	}
```
[the remark this answers is not in the document]
Comment (NextG3nMxJ) [reply pkg/llm/anthropic.go]: Fixed in 56f9dd5, with one adjustment: custom `BaseURL` must be `https`, except loopback `http` (`127.0.0.1` / `localhost` / `::1`) so unit tests can keep using `httptest`. Production default remains `https://api.anthropic.com`.

File: pkg/llm/anthropic.go [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+	err error
+}
+
+func (r *retryableError) Error() string { return r.err.Error() }
+func (r *retryableError) Unwrap() error { return r.err }
+
+func (c *anthropicClient) Complete(ctx context.Context, req *Request) (*Response, error) {
+	msgs := make([]anthropicMessage, len(req.Messages))
+	for i, m := range req.Messages {
+		msgs[i] = anthropicMessage{Role: m.Role, Content: m.Content}
```
[the remark this answers is not in the document]
Comment (NextG3nMxJ) [reply pkg/llm/anthropic.go]: Fixed in 56f9dd5. Conversion is now `anthropicMessage(m)`.

File: pkg/llm/anthropic.go [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+		if delay > 0 {
+			t := time.NewTimer(delay)
+			defer t.Stop()
+			select {
+			case <-t.C:
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			}
+		}
+		return nil, &retryableError{err: fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(respBody))}
```
[the remark this answers is not in the document]
Comment (NextG3nMxJ) [reply pkg/llm/anthropic.go]: Fixed in 56f9dd5. `doRequest` returns retry metadata without sleeping; `Complete` delays only when `attempt < retries`.

File: pkg/llm/client.go [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+	MaxRetries int
+	BaseURL    string
+}
+
+func WithTimeout(d time.Duration) Option {
+	return func(c *clientConfig) { c.Timeout = d }
+}
+
+func WithMaxRetries(n int) Option {
+	return func(c *clientConfig) { c.MaxRetries = n }
```
[the remark this answers is not in the document]
Comment (NextG3nMxJ) [reply pkg/llm/client.go]: Fixed in 56f9dd5. `WithMaxRetries` clamps negative values to 0.

File: pkg/llm/sanitize.go [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+// Sanitize strips ANSI escape sequences and control characters (including
+// null bytes) from input, while preserving newlines, carriage returns, and
+// tabs. It is used to clean untrusted text before including it in an LLM
+// prompt.
+func Sanitize(input string) string {
+	input = ansiPattern.ReplaceAllString(input, "")
+	var b strings.Builder
+	b.Grow(len(input))
+	for _, r := range input {
+		if r == '\n' || r == '\r' || r == '\t' || r >= 0x20 {
```
[the remark this answers is not in the document]
Comment (NextG3nMxJ) [reply pkg/llm/sanitize.go]: Fixed in 56f9dd5. `WrapUntrusted` HTML-escapes after `Sanitize`, and there is a regression test for an embedded closing tag.

File: pkg/scoring/go_scorers.go [hunk 21]
```diff
@@ -26,8 +28,13 @@ func BuiltinGoScorers() []*Scorer {
...
 // identical — assembling the list in only one path silently drops Go-scored
 // rules (e.g. classic GitHub PATs) from the other.
-func AllBuiltinScorers() ([]*Scorer, error) {
-	yamlScorers, err := NewLoader().LoadBuiltinScorers()
+//
+// llmClient is threaded through to every YAML scorer's llm: conditions. Pass
+// nil when LLM scoring is unavailable (e.g. --score-scope disabled or no
+// TITUS_LLM_API_KEY) — llm: conditions with a nil client silently evaluate
+// to false rather than firing.
+func AllBuiltinScorers(llmClient llm.Client) ([]*Scorer, error) {
```
[the remark this answers is not in the document]
Comment (NextG3nMxJ) [reply pkg/scoring/go_scorers.go]: Fixed in 56f9dd5. `AllBuiltinScorers` is now variadic (`llmClient ...llm.Client`), so existing zero-argument callers still compile. Pass a client only when LLM scoring is configured.

File: pkg/validator/jenkins.go [hunk 22]
```diff
@@ -76,39 +76,46 @@ func (v *JenkinsValidator) Validate(ctx context.Context, match *types.Match) (*t
 	}
 	defer func() { _ = resp.Body.Close() }()
 
+	body, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<16))
```
[the remark this answers is not in the document]
Comment (NextG3nMxJ) [reply pkg/validator/jenkins.go]: Fixed in 56f9dd5. Jenkins now handles the `ReadAll` error and rejects bodies larger than 64 KiB instead of parsing a truncated payload.

File: pkg/validator/retry.go [hunk 23]
```diff
@@ -0,0 +1,152 @@
...
+	retryBackoffDelay = 1 * time.Second
+	retryAfterCap     = 30 * time.Second
+)
+
+// RetryHTTPClient wraps an *http.Client with single-retry logic for
+// transient failures: 429 (respecting Retry-After, capped), 5xx responses,
+// and connection errors. It is a drop-in replacement for *http.Client.Do.
+// An optional BackoffController records errors/successes and is consulted
+// before each attempt to apply adaptive backoff across concurrent callers.
+type RetryHTTPClient struct {
```
[the remark this answers is not in the document]
Comment (NextG3nMxJ) [reply pkg/validator/retry.go:43]: Addressed in 56f9dd5 without changing `HTTPValidator` to a custom `Do` interface. `SetBackoff` installs `RetryHTTPClient` as the `*http.Client` Transport via `WrapHTTPClient`, so timeout/redirect/cookie behavior stays on the outer client. YAML HTTP validators and the Go validators that own an `*http.Client` are wrapped.

File: pkg/validator/retry.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (NextG3nMxJ) [reply pkg/validator/retry.go]: Fixed in 56f9dd5. Retries rewind through `req.GetBody` and skip replay when no replayable body exists. Added a POST-body retry test.

File: pkg/validator/verifier.go [hunk 24]
```diff
@@ -0,0 +1,259 @@
...
+// (transport error or unparseable response) falls back to original.
+func (v *LLMVerifier) tryLLMUpgrade(ctx context.Context, match *types.Match, original *types.ValidationResult) *types.ValidationResult {
+	userMsg := v.buildUserMessage(match, original)
+
+	key := llm.CacheKey(verifierSystemPrompt, userMsg)
+	if cached := v.cache.Get(key); cached != nil {
+		return v.applyLLMResponse(cached.Content, original)
+	}
+
+	v.spent.Add(1)
```
[the remark this answers is not in the document]
Comment (NextG3nMxJ) [reply pkg/validator/verifier.go]: Fixed in 56f9dd5. Reservation is a CAS loop immediately before `Complete`, after the cache check. Callers that lose the race or hit the cap keep the original result.

File: pkg/validator/verifier.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (NextG3nMxJ) [reply pkg/validator/verifier.go]: Fixed in 56f9dd5. Confidence outside `[0, 1]` is rejected, upgrades require `>= 0.7`, and an existing valid/invalid confidence is not lowered.

File: docs/superpowers/plans/2026-09-14-agentic-validation-pipeline.md
Comment (michaelweber) [inline docs/superpowers/plans/2026-09-14-agentic-validation-pipeline.md:1]: We don't need to check this in.

File: docs/superpowers/specs/2026-09-14-agentic-validation-pipeline-design.md
Comment (michaelweber) [inline docs/superpowers/specs/2026-09-14-agentic-validation-pipeline-design.md:1]: We don't need to check this in either.

**Review by michaelweber — CHANGES_REQUESTED** [review]
I think we don't need an LLM to evaluate edge cases for whether or not something validates. But we do need a mechanism to make it easy for something to agentically call titus. The goal is to avoid drift from our CLI and any agent running this.

File: docs/superpowers/plans/2026-09-14-agentic-validation-pipeline.md
Comment (NextG3nMxJ) [reply docs/superpowers/plans/2026-09-14-agentic-validation-pipeline.md:1]: Removed in 87f20e2. Spec/plan docs will not be checked in on this PR.

File: docs/superpowers/specs/2026-09-14-agentic-validation-pipeline-design.md
Comment (NextG3nMxJ) [reply docs/superpowers/specs/2026-09-14-agentic-validation-pipeline-design.md:1]: Removed in 87f20e2. Spec/plan docs will not be checked in on this PR.

**Comment by NextG3nMxJ** [comment]
Dropped the LLM validator/scoring path in 87f20e2.

Titus does not call an LLM for validation or scoring. Those stay programmatic. This PR is now backoff, HTTP retry, and `--validate-rate-limit` only. Spec/plan files are gone.

CLI remains the interface for anything (including an agent) that needs to run titus — that is separate from putting an LLM inside the validator.

**Comment by NextG3nMxJ** [comment]
Closing without merge.

Titus already covers what Marcus needs without an in-process LLM: identify via `scan` / the Go library, validate via `--validate` / `analyze`, score via the scoring engine. Guard already calls Titus from `titus-agent` → `secrets` capability. The LLM verifier/scoring path was the wrong assignment, and the leftover retry/rate-limit work is not required for that.

Branch will be deleted.
